### PR TITLE
Adding access restriction feature

### PR DESCRIPTION
**Pull request by willyjlee**

### CODE REVIEW

File: src/codeu/chat/client/core/Controller.java [hunk 1]
```diff
@@ -39,6 +39,80 @@ public Controller(ConnectionSource source) {
     this.source = source;
   }
 
+  // TODO: add return options, no user found, no access,
```
Comment (adrnswanberg) [inline src/codeu/chat/client/core/Controller.java]: Implement this TODO or remove it if it does not necessary to complete this feature.

File: src/codeu/chat/client/core/Controller.java [hunk 2]
```diff
@@ -39,6 +39,80 @@ public Controller(ConnectionSource source) {
...
+      }
+    } catch (Exception ex) {
+      System.out.println("ERROR: Exception during call on server. Check log for details.");
+      LOG.error(ex, "Exception during call on server.");
+      return false;
+    }
+  }
+
+  public String getAccess(Uuid conversation, Uuid user) {
+
```
Comment (adrnswanberg) [inline src/codeu/chat/client/core/Controller.java]: nit: remove blank line

File: src/codeu/chat/client/core/Controller.java [hunk 3]
```diff
@@ -39,6 +39,80 @@ public Controller(ConnectionSource source) {
...
+      }
+    } catch (Exception ex) {
+      System.out.println("ERROR: Exception during call on server. Check log for details.");
+      LOG.error(ex, "Exception during call on server.");
+    }
+    return response;
+  }
+
+  public boolean joinConversation(Uuid conversation, Uuid user) {
+
```
Comment (adrnswanberg) [inline src/codeu/chat/client/core/Controller.java]: same here

File: src/codeu/chat/client/core/ConversationContext.java [hunk 4]
```diff
@@ -32,19 +32,29 @@
...
 
     this.user = user;
     this.conversation = conversation;
     this.view = view;
     this.controller = controller;
   }
 
+  // access = "member", "owner", "remove"
+  // return if ok
+  public boolean changeAccess(Uuid requestor, String userName, String access, Uuid conversation) {
```
Comment (adrnswanberg) [inline src/codeu/chat/client/core/ConversationContext.java]: If access really only has three allowed values, you should probably make it an enum.

File: src/codeu/chat/common/AccessCode.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package codeu.chat.common;
+
+
+public final class AccessCode {
```
Comment (adrnswanberg) [inline src/codeu/chat/common/AccessCode.java]: This seems like a duplicate of Access.java? If so, remove and use Access.java everywhere.

File: src/codeu/chat/common/Access.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
+package codeu.chat.common;
+
+public enum Access {
```
Comment (adrnswanberg) [inline src/codeu/chat/common/Access.java]: This doesn't conform to the google style guide: https://google.github.io/styleguide/javaguide.html#s4.8.1-enum-classes

I'm skeptical that you need the extra methods here too. I know java lets you do a lot with enum classes, but usually you just need a few definitions and that's it.

File: src/codeu/chat/common/ChangeAccessRequest.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+  public void write(OutputStream out, Object value) throws IOException {
+    SERIALIZER.write(out, (ChangeAccessRequest)value);
+  }
+
+  // map from conversation id to Access
+  public final Uuid user;
+  public final Uuid conversation;
+  public final Access access;
+
+  public ChangeAccessRequest(Uuid user, Uuid conversation, Access access) {
```
Comment (adrnswanberg) [inline src/codeu/chat/common/ChangeAccessRequest.java]: nit: prefer to keep instance variables and constructor at the top of the class (in that order).

File: src/codeu/chat/common/ChangeAccessRequest.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+    this.user = user;
+    this.conversation = conversation;
+    this.access = access;
+  }
+
+  public static final Serializer<ChangeAccessRequest> SERIALIZER = new Serializer<ChangeAccessRequest>() {
+
```
Comment (adrnswanberg) [inline src/codeu/chat/common/ChangeAccessRequest.java]: nit: remove blank line

File: src/codeu/chat/common/ChangeAccessRequest.java [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+      Uuid.SERIALIZER.write(out, value.user);
+      Uuid.SERIALIZER.write(out, value.conversation);
+      Serializers.INTEGER.write(out, value.access.getAccessNum());
+    }
+
+    @Override
+    public ChangeAccessRequest read(InputStream in) throws IOException {
+
+      return new ChangeAccessRequest(Uuid.SERIALIZER.read(in), Uuid.SERIALIZER.read(in), Access.get(Serializers.INTEGER.read(in)));
+      // TODO: add reading convoid, access, convoid, access, ...
```
Comment (adrnswanberg) [inline src/codeu/chat/common/ChangeAccessRequest.java]: implement this todo or remove it if it is no longer necessary. also remove commented out code.

File: src/codeu/chat/common/ChangeAccessRequest.java [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    @Override
+    public void write(OutputStream out, ChangeAccessRequest value) throws IOException {
+      Uuid.SERIALIZER.write(out, value.user);
+      Uuid.SERIALIZER.write(out, value.conversation);
+      Serializers.INTEGER.write(out, value.access.getAccessNum());
+    }
+
+    @Override
+    public ChangeAccessRequest read(InputStream in) throws IOException {
+
```
Comment (adrnswanberg) [inline src/codeu/chat/common/ChangeAccessRequest.java]: nit: remove blank line

File: src/codeu/chat/common/ConversationHeader.java [hunk 11]
```diff
@@ -64,6 +64,19 @@ public ConversationHeader read(InputStream in) throws IOException {
     }
   };
 
+    // @Override
```
Comment (adrnswanberg) [inline src/codeu/chat/common/ConversationHeader.java]: remove commented out code.

File: src/codeu/chat/common/User.java [hunk 12]
```diff
@@ -62,22 +64,41 @@ public User read(InputStream in) throws IOException {
...
+  public Access get(Uuid conversation) {
+    return map.get(conversation);
+  }
+
   public final Uuid id;
   public final String name;
   public final Time creation;
 
+  // map from conversation id to Access
+  private Map<Uuid, Access> map;
```
Comment (adrnswanberg) [inline src/codeu/chat/common/User.java]: give this a more descriptive name, and move it closer to the other instance variables.

File: src/codeu/chat/common/Writeable.java [hunk 13]
```diff
@@ -14,6 +14,7 @@
   public static final String USER_STR = "user";
   public static final String MESSAGE_STR = "message";
   public static final String CONVERSATION_STR = "conversationheader";
+  public static final String CHANGE_ACCESS_REQUEST_STR = "changeaccessrequest";
```
Comment (adrnswanberg) [inline src/codeu/chat/common/Writeable.java:17]: in retrospect, these could have probably been enum values. you don't necessarily have to change it though.

File: src/codeu/chat/common/User.java [hunk 14]
```diff
@@ -62,22 +64,41 @@ public User read(InputStream in) throws IOException {
...
           Time.SERIALIZER.read(in)
       );
-
     }
   };
   
   public HashMap<Uuid, Time> UserUpdateMap = new HashMap<Uuid, Time>();
   public HashMap<Uuid, Time> ConvoUpdateMap = new HashMap<Uuid, Time>();
 
+  public void add(Uuid conversation, Access access) {
```
Comment (adrnswanberg) [inline src/codeu/chat/common/User.java]: These methods should have more descriptive names.

File: src/codeu/chat/server/Controller.java [hunk 15]
```diff
@@ -68,6 +72,88 @@ private void save(Writeable x) {
...
+
+    User requestorUser = model.userById().first(requestor);
+    User user = model.userByText().first(userName);
+
+    if(requestorUser == null || user == null) {
+      return false;
+    }
+
+    Access requestorAccess = requestorUser.get(conversation);
+    if (requestorAccess!=Access.CREATOR && requestorAccess!=Access.OWNER) {
```
Comment (adrnswanberg) [inline src/codeu/chat/server/Controller.java]: nit: add space between '!='

File: src/codeu/chat/server/Controller.java [hunk 16]
```diff
@@ -68,6 +72,88 @@ private void save(Writeable x) {
...
+      // trying to change self
+      return false;
+    }
+
+    // can't change a CREATOR
+    if (user.get(conversation) == Access.CREATOR) {
+      return false;
+    }
+
+    if (access.equals(AccessCode.CREATOR)) {
```
Comment (adrnswanberg) [inline src/codeu/chat/server/Controller.java]: This should be up higher before other checks.

File: src/codeu/chat/server/Controller.java [hunk 17]
```diff
@@ -68,6 +72,88 @@ private void save(Writeable x) {
...
+      return false;
+    }
+
+    // TODO: add persistent here
```
Comment (adrnswanberg) [inline src/codeu/chat/server/Controller.java]: Implement TODO or remove if it is no longer necessary.

File: src/codeu/chat/server/Controller.java [hunk 18]
```diff
@@ -68,6 +72,88 @@ private void save(Writeable x) {
...
+    User user = model.userById().first(userid);
+    ConversationHeader conversation = model.conversationById().first(conversationid);
+
+    if (access == Access.NO_ACCESS) {
+      user.remove(conversationid);
+    } else {
+      user.add(conversationid, access);
+    }
+    LOG.info("loadChangeAccess success: (user=%s, conversation=%s, access=%s)", user.name, conversation.title, access);
+
```
Comment (adrnswanberg) [inline src/codeu/chat/server/Controller.java]: nit: remove blank line

File: src/codeu/chat/server/Controller.java [hunk 19]
```diff
@@ -68,6 +72,88 @@ private void save(Writeable x) {
...
+        save(new ChangeAccessRequest(user.id, conversation, Access.NO_ACCESS));
+        break;
+      default:
+        return false;
+    }
+    return true;
+  }
+
+  public void loadChangeAccess(Uuid userid, Uuid conversationid, Access access) {
+
```
Comment (adrnswanberg) [inline src/codeu/chat/server/Controller.java]: nit: remove blank line

File: src/codeu/chat/server/Controller.java [hunk 20]
```diff
@@ -68,6 +72,88 @@ private void save(Writeable x) {
...
+      case CREATOR: return AccessCode.CREATOR;
+      default: return AccessCode.NO_ACCESS;
+    }
+  }
+
+  public boolean joinConversation(Uuid conversation, Uuid user) {
+    User getUser = model.userById().first(user);
+
+    return getUser.containsConversation(conversation);
+    //getUser.add(conversation, Access.MEMBER);
```
Comment (adrnswanberg) [inline src/codeu/chat/server/Controller.java]: remove commented out code, and blank line below.

File: src/codeu/chat/server/Controller.java [hunk 21]
```diff
@@ -137,7 +223,15 @@ public User newUser(Uuid id, String name, Time creationTime) {
 
     User user = null;
 
-    if (isIdFree(id)) {
+    if (isUsernameInUse(name)) {
+
```
Comment (adrnswanberg) [inline src/codeu/chat/server/Controller.java]: nit: remove blank line

File: src/codeu/chat/server/Controller.java [hunk 22]
```diff
@@ -137,7 +223,15 @@ public User newUser(Uuid id, String name, Time creationTime) {
...
+      LOG.info(
+          "newUser fail - name in use (user.id=%s user.name=%s user.time=%s)",
+          id,
+          name,
+          creationTime);
+
```
Comment (adrnswanberg) [inline src/codeu/chat/server/Controller.java]: same here

File: src/codeu/chat/util/Time.java [hunk 23]
```diff
@@ -48,6 +48,18 @@ public Time read(InputStream in) throws IOException {
 
   public long inMs() { return date.getTime(); }
 
+  // @Override
```
Comment (adrnswanberg) [inline src/codeu/chat/util/Time.java]: remove commented out code

File: src/codeu/chat/common/AccessCode.java (hunk 5 above)
Comment (adrnswanberg) [reply src/codeu/chat/common/AccessCode.java]: Yeah, so after looking at the rest of the changes, you can definitely get rid of this and just use Access everywhere.

**Review by adrnswanberg — CHANGES_REQUESTED** [review]
Nice work! Nothing too major from me. Address my comments and you'll be good to merge. Let's push hard and try to get everything merged by Friday!

File: src/codeu/chat/common/AccessCode.java [hunk 24]
```diff
@@ -1,25 +1,25 @@
-// Copyright 2017 Google Inc.
+// // Copyright 2017 Google Inc.
```
Comment (adrnswanberg) [inline src/codeu/chat/common/AccessCode.java]: delete this file

**Review by adrnswanberg — APPROVED** [review]
Nice, that's a wrap! You should all be proud of yourselves :)

File: src/codeu/chat/common/ChangeAccessRequest.java (hunk 7 above)
Comment (kaki123) [reply src/codeu/chat/common/ChangeAccessRequest.java]: This is resolved

**Pull request merged**